[RFC] Add OperatorHandle overload to the RecordFunction::before() method

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46401 [RFC] Add OperatorHandle overload to the RecordFunction::before() method**

Broader context about selective/custom build available at https://fb.quip.com/2oEzAR5MKqbD and https://fb.workplace.com/groups/pytorch.mobile.team/permalink/735794523641956/

Basically, we want to be able to trace full operator names (with overload name). The current observer infra picks up the operator name from the schema, which doesn't seem to include the overload name. To ensure consistency with the existing uses and to accomodate the new use-case, this diff adds a new overload to accept an `OperatorHandle` object, and the code in `before()` eagerly resolves it to an `OperatorName` object (which can be cached in a member variable) as well as a string (view) operator-name which has the same semantics as before.

Why do we pass in an `OperatorHandle` but then resolve it to an `OperatorName`? This might come across as a strange design choice (and it is), but it is grounded in practicality.

It is not reasonable to cache an `OperatorHandle` object but caching an `OperatorName` object is reasonable since it holds all the data itself.

An initial version of this change was trying to test this change in the `xplat` repo, which didn't work. Thanks to @iliacher for pointing out that the dispatcher observing mechanism is disabled under a compile time flag (macro) for xplat.

Differential Revision: [D24315241](https://our.internmc.facebook.com/intern/diff/D24315241/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24315241/)!

dhruvbird committed:
[RFC] Add OperatorHandle overload to the RecordFunction::before() method

Broader context about selective/custom build available at https://fb.quip.com/2oEzAR5MKqbD and https://fb.workplace.com/groups/pytorch.mobile.team/permalink/735794523641956/

Basically, we want to be able to trace full operator names (with overload name). The current observer infra picks up the operator name from the schema, which doesn't seem to include the overload name. To ensure consistency with the existing uses and to accomodate the new use-case, this diff adds a new overload to accept an `OperatorHandle` object, and the code in `before()` eagerly resolves it to an `OperatorName` object (which can be cached in a member variable) as well as a string (view) operator-name which has the same semantics as before.

Why do we pass in an `OperatorHandle` but then resolve it to an `OperatorName`? This might come across as a strange design choice (and it is), but it is grounded in practicality.

It is not reasonable to cache an `OperatorHandle` object but caching an `OperatorName` object is reasonable since it holds all the data itself.

An initial version of this change was trying to test this change in the `xplat` repo, which didn't work. Thanks to @iliacher for pointing out that the dispatcher observing mechanism is disabled under a compile time flag (macro) for xplat.

Differential Revision: [D24315241](https://our.internmc.facebook.com/intern/diff/D24315241/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24315241/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Add OperatorHandle overload to the RecordFunction::before() method"


Broader context about selective/custom build available at https://fb.quip.com/2oEzAR5MKqbD and https://fb.workplace.com/groups/pytorch.mobile.team/permalink/735794523641956/

Basically, we want to be able to trace full operator names (with overload name). The current observer infra picks up the operator name from the schema, which doesn't seem to include the overload name. To ensure consistency with the existing uses and to accomodate the new use-case, this diff adds a new overload to accept an `OperatorHandle` object, and the code in `before()` eagerly resolves it to an `OperatorName` object (which can be cached in a member variable) as well as a string (view) operator-name which has the same semantics as before.

Why do we pass in an `OperatorHandle` but then resolve it to an `OperatorName`? This might come across as a strange design choice (and it is), but it is grounded in practicality.

It is not reasonable to cache an `OperatorHandle` object but caching an `OperatorName` object is reasonable since it holds all the data itself.

An initial version of this change was trying to test this change in the `xplat` repo, which didn't work. Thanks to @iliacher for pointing out that the dispatcher observing mechanism is disabled under a compile time flag (macro) for xplat.

Differential Revision: [D24315241](https://our.internmc.facebook.com/intern/diff/D24315241/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24315241/)!

[ghstack-poisoned]

ilia-cher reviewed (commented):
let's run a benchmark, specifically could you first run benchmarks/profiler_benchmark/profiler_bench.py on the PR and on base (the interesting use case is when profiler is not enabled)

dhruvbird commented on aten/src/ATen/record_function.h:
```diff
@@ -1,12 +1,18 @@
 
 #include <functional>
 
+namespace c10 {
+class CAFFE2_API OperatorHandle;
+}
```
The overload below `void before(c10::OperatorHandle const& op, int64_t sequence_nr = -1);` takes in `c10:OperatorHandle const&`, and the class doesn't seem to be forward declared. Also I think OperatorHandle itself is declared in Dispatcher.h, so including it would cause a circular dependency.

dhruvbird commented:
> benchmarks/profiler_benchmark/profiler_bench.py

Thanks! Here's are the benchmark results: https://www.internalfb.com/intern/paste/P146511280/

Seems like there's no difference as such.

dhruvbird committed:
Update on "[RFC] Add OperatorHandle overload to the RecordFunction::before() method"


Broader context about selective/custom build available at https://fb.quip.com/2oEzAR5MKqbD and https://fb.workplace.com/groups/pytorch.mobile.team/permalink/735794523641956/

Basically, we want to be able to trace full operator names (with overload name). The current observer infra picks up the operator name from the schema, which doesn't seem to include the overload name. To ensure consistency with the existing uses and to accomodate the new use-case, this diff adds a new overload to accept an `OperatorHandle` object, and the code in `before()` eagerly resolves it to an `OperatorName` object (which can be cached in a member variable) as well as a string (view) operator-name which has the same semantics as before.

Why do we pass in an `OperatorHandle` but then resolve it to an `OperatorName`? This might come across as a strange design choice (and it is), but it is grounded in practicality.

It is not reasonable to cache an `OperatorHandle` object but caching an `OperatorName` object is reasonable since it holds all the data itself.

An initial version of this change was trying to test this change in the `xplat` repo, which didn't work. Thanks to @iliacher for pointing out that the dispatcher observing mechanism is disabled under a compile time flag (macro) for xplat.

Differential Revision: [D24315241](https://our.internmc.facebook.com/intern/diff/D24315241/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24315241/)!

[ghstack-poisoned]

ilia-cher commented:
thanks!, could you use the following snippet to run instruction counts on the base rev. and with this PR:
```
import torch
from torch.utils.benchmark import Timer
import timeit

counts = Timer(
    "x.backward()",
    setup="x = torch.ones((1,)) + torch.ones((1,), requires_grad=True)",
    timer=timeit.default_timer,
).collect_callgrind()

print(counts)
```

dhruvbird commented:
> thanks!, could you use the following snippet to run instruction counts on the base rev. and with this PR:
> 
> ```
> import torch
> from torch.utils.benchmark import Timer
> import timeit
> 
> counts = Timer(
>     "x.backward()",
>     setup="x = torch.ones((1,)) + torch.ones((1,), requires_grad=True)",
>     timer=timeit.default_timer,
> ).collect_callgrind()
> 
> print(counts)
> ```

Here are the results: https://www.internalfb.com/intern/paste/P146560623/ and pastebin link: https://pastebin.com/QtgkZqpm

@ilia-cher Please could you help me understand how to interpret the results?

ilia-cher commented:
> > thanks!, could you use the following snippet to run instruction counts on the base rev. and with this PR:
> > ```
> > import torch
> > from torch.utils.benchmark import Timer
> > import timeit
> > 
> > counts = Timer(
> >     "x.backward()",
> >     setup="x = torch.ones((1,)) + torch.ones((1,), requires_grad=True)",
> >     timer=timeit.default_timer,
> > ).collect_callgrind()
> > 
> > print(counts)
> > ```
> 
> Here are the results: https://www.internalfb.com/intern/paste/P146560623/ and pastebin link: https://pastebin.com/QtgkZqpm
> 
> @ilia-cher Please could you help me understand how to interpret the results?

when time difference is very small to measure, we're mainly interested in number of instructions (noisy symbols removed) - lgtm

ilia-cher commented on aten/src/ATen/record_function.cpp:
```diff
@@ -387,6 +389,21 @@ void RecordFunction::before(std::string name, int64_t sequence_nr) {
+  }
+  sequence_nr_ = sequence_nr;
+  thread_id_ = currentThreadId();
+  operator_name_ = op.operator_name();
+  name_ = StringView(op.schema().name());
```
seems like we could've store just OperatorHandle? if we want more from it I guess we could refactor it later

ilia-cher reviewed (approved):
LG, thanks!

robieta commented:
> > > thanks!, could you use the following snippet to run instruction counts on the base rev. and with this PR:
> > > ```
> > > import torch
> > > from torch.utils.benchmark import Timer
> > > import timeit
> > > 
> > > counts = Timer(
> > >     "x.backward()",
> > >     setup="x = torch.ones((1,)) + torch.ones((1,), requires_grad=True)",
> > >     timer=timeit.default_timer,
> > > ).collect_callgrind()
> > > 
> > > print(counts)
> > > ```
> > 
> > 
> > Here are the results: https://www.internalfb.com/intern/paste/P146560623/ and pastebin link: https://pastebin.com/QtgkZqpm
> > @ilia-cher Please could you help me understand how to interpret the results?
> 
> when time difference is very small to measure, we're mainly interested in number of instructions (noisy symbols removed) - lgtm (assuming some noise in the output)

One thing that can help is to actually look at the instructions that are different to get a sense of what is changing in addition to just the overall number. (If you haven't seen my recent talk, that should get you started. And docs are coming.)

dhruvbird commented:
> One thing that can help is to actually look at the instructions that are different to get a sense of what is changing in addition to just the overall number. (If you haven't seen my recent talk, that should get you started. And docs are coming.)

+1 -- will be looking out for how - I'm assuming you'll have some pre-created benchmarks checked in that I can run on my change set(s).

robieta commented:
> > One thing that can help is to actually look at the instructions that are different to get a sense of what is changing in addition to just the overall number. (If you haven't seen my recent talk, that should get you started. And docs are coming.)
> 
> +1 -- will be looking out for how - I'm assuming you'll have some pre-created benchmarks checked in that I can run on my change set(s).

At some point, though not currently.
